### PR TITLE
[wip] implement client dashboard api for notes

### PR DESCRIPTION
**Pull request by icewind1991**
Dashboard api will allow clients to render the dashboard widgets

See https://github.com/nextcloud/server/pull/33658 for more information, requires the interfaces added by that pr

This implementation was created to provide an example for a dashboard that has a button. Interfaces might still be subject to change.

### CODE REVIEW

File: lib/AppInfo/DashboardWidget.php [hunk 1]
```diff
@@ -61,4 +69,31 @@ public function getUrl(): ?string {
...
+	public function getButtonUrl(): string {
+		return $this->url->linkToRouteAbsolute('notes.page.create');
+	}
+
+	public function getButtonIconUrl(): ?string {
+		return null;
+	}
+
+	public function getButtonText(): string {
+		return $this->l10n->t('New notes');
```
Comment (stefan-niedermann) [inline lib/AppInfo/DashboardWidget.php:82]: Shouldn't this be singular? <kbd>New note</kbd>?

File: lib/AppInfo/DashboardWidget.php [hunk 2]
```diff
@@ -61,4 +70,29 @@ public function getUrl(): ?string {
...
+	public function getItems(string $userId, ?string $since = null, int $limit = 7): array {
+		$notes = $this->notesService->getTopNotes($userId);
+		$notes = array_slice($notes, 0, $limit);
+		return array_map(function (Note $note) {
+			$excerpt = '';
+			try {
+				$excerpt = $note->getExcerpt();
+			} catch (\Throwable $e) {
+			}
+			$link = $this->url->linkToRouteAbsolute('notes.page.index', ['id' => $note->getId()]);
```
Comment (tobiasKaminsky) [inline lib/AppInfo/DashboardWidget.php:93]: Is this correct link?
It is something like: https://cloud.com/index.php/apps/notes/?id=271
but on web I see:
https://cloud.com/apps/notes/note/6004520

Thus all links via widget end up showing any previous note but not the desired one.

Comment (korelstar) [reply lib/AppInfo/DashboardWidget.php:93]: Yep, it must be `notes.page.indexnote` instead:

```suggestion
			$link = $this->url->linkToRouteAbsolute('notes.page.indexnote', ['id' => $note->getId()]);
```

See e.g. https://github.com/nextcloud/notes/blob/v4.5.0/lib/Controller/PageController.php#L62

**Comment by korelstar** [comment]
Nice! How can I test this (i.e. which clients will use that API)? Is it for Nextcloud Android/iOS or for the Nextcloud server itself? The last one would be interesting, too, because currently, the dashboard app loads several megabytes Javascript for every widget.

**Comment by tobiasKaminsky** [comment]
It can be used by any client, but first we will implement it on Android/iOS.

**Comment by tobiasKaminsky** [comment]
Bug: iconURL is empty although it is shown in web @robi

**Comment by tobiasKaminsky** [comment]
Bug: notes icon url on widgets-items endpoint is empty

**Comment by juliushaertl** [comment]
@tobiasKaminsky I rebased and pushed a commit to fix the icon bugs. Mind to give this another test run?

**Comment by tobiasKaminsky** [comment]
In {{ base_url  }}/ocs/v2.php/apps/dashboard/api/v1/widgets:
icon_url: "\/nc\/apps\/notes\/img\/notes-dark.svg"

but it must be absolute, otherwise clients cannot use it.


In: {{ base_url  }}/ocs/v2.php/apps/dashboard/api/v1/widget-items
iconURL: is starred.svg or notes-dark.svg
But it should provide preview of text.

image

**Pull request merged**